File: Problems 1900-1999/1930. Unique Length-3 Palindromic Subsequences/solution.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        // palindrome of length 3:
        // 1. all same character: aaa
        // 2. one different character in middle: aba
        // --> we need to know how many
        // unique characters come after current char and
        // last occurence of current char

        vector<int> first(26, -1), last(26, -1);

        for (int i = 0; i < s.length(); i++) {
            if (first[s[i]-'a'] == -1) first[s[i]-'a'] = i;
            last[s[i]-'a'] = i;
        }

        int res = 0;
        for (int i = 0; i < 26; i++) {
            if (first[i] < last[i]) {
                res += unordered_set<char>(s.begin() + first[i] + 1, s.begin() + last[i]).size();
            }
        }
        return res;
    }
};

File: Problems 1900-1999/1930. Unique Length-3 Palindromic Subsequences/solution.java
class Solution {
    public int countPalindromicSubsequence(String s) {
        int[] first = new int[26];
        int[] last = new int[26];

        int n = s.length();

        for (int i = 0; i < 26; i++) {
            first[i] = -1;
            last[i] = n;
        }

        for (int i = 0; i < n; i++) {
            int c = s.charAt(i) - 'a';
            if (first[c] == -1) first[c] = i;
            last[c] = i;
        }

        int res = 0;
        for (int i = 0; i < 26; i++) {
            if (first[i] > -1 && last[i] < n) {
                res += count(first[i], last[i], s);
            }
        }

        return res;
    }

    private int count(int start, int end, String s) {
        if (start >= end) return 0;
        int[] count = new int[26];

        for (int i = start + 1; i <= end - 1; i++) {
            count[s.charAt(i) - 'a']++;
        }

        int res = 0;
        for (int freq : count) {
            if (freq > 0) res++;
        }
        return res;
    }
}